lib/SILOptimizer/Transforms/AccessEnforcementDom.cpp

```diff
@@ -65,417 +57,549 @@
+  // If the previous domNode does not dominate currDomNode, then replace it.
+  DominatingAccess &domAccess = iterAndInserted.first->second;
+  if (!domInfo->dominates(domAccess.domNode, currDomNode)) {
+    domAccess = DominatingAccess(BAI, currDomNode);
+    return false;
   }
+  // The previously mapped access still dominates this block, so the current
+  // access can potentially be optimized.
+  return optimizeDominatedAccess(BAI, currDomStorage, domAccess);
 }
 
-void LoopDominatingAccessAdder::perform() {
-  // Process loops bottom up in the loop tree.
-  while (!bottomUpWorkList.empty()) {
-    SILLoop *currnetLoop = bottomUpWorkList.pop_back_val();
-    LLVM_DEBUG(llvm::dbgs() << "Processing loop " << *currnetLoop);
+// If possible, optimize the current access by converting it to [static]. Return
+// true if the optimization succeeds.
+//
+// This function is not allowed to add or erase instructions, only change
+// instruction flags.
+//
+// The four required conditions for converting this access to static are:
+//
+// 1. A closed dominating access has identical storage.
+//
+// The caller looked up this access' storage in storageToDomMap and checked
+// that the previously seen access dominates this block. As long as this access'
+// isInner flag from DominatedAccessAnalysis is not set, the dominating access
+// must be closed.
+//
+// 2. There is no open (overlapping) access with nondistinct storage (that isn't
+// also open at the dominating access).
+//
+// The isInner flag from DominatedAccessAnalysis indicates no enclosing
+// nondistinct scopes within this function. Any outer scope would enclose the
+// whole function, thereby also enclosing the dominating scope.
+//
+// 3. This access has no_nested_conflict.
+//
+// This is a direct check on this BeginAccessInst flag.
+//
+// 4. The current and dominating access kinds are compatible:
+//
+// read -> read: OK
+//
+// modify -> read: OK
+//
+// modify -> modify: OK
+//
+// read -> modify: Requires promoting the dominating access to a modify. This
+// can be done as long as the dominating access does not contain another
+// non-distinct read and isn't contained by another non-distinct read. The
+// containsRead and isInner flags from in DominatedAccessAnalysis answer this
+// conservatively.
+bool DominatedAccessRemoval::optimizeDominatedAccess(
+    BeginAccessInst *BAI, DomAccessedStorage currAccessInfo,
+    const DominatingAccess &domAccess) {
+  // 1. and 2. If any nondistinct scopes are open, it must remain dynamic.
+  if (currAccessInfo.isInner())
+    return false;
+
+  // 3. If BAI may have a nested conflict, it must remain dynamic.
+  if (!BAI->hasNoNestedConflict())
+    return false;
 
-    llvm::SmallVector<BeginAccessInst *, 8> dominatableInstrVec;
-    analyzeCurrentLoop(currnetLoop, dominatableInstrVec);
-    optimizeLoop(currnetLoop, dominatableInstrVec);
+  // 4. Promoting a read to a modify is only safe with no nested reads.
```

Comment:
I am not sure this is safe: imagine a dominating read in the prolog, and a dominated modify that's part of a conditional control flow that is 'dead code' at runtime, with a read access at the function's call site, we might introduce a false positive?

Reply:
@shajrawi Thanks for the review. You raise an excellent point. I added comments here explaining why I consider it a valid transformation.

```
// Note: Promoting an earlier access to a modify could cause a program to
// trap when optimized even if the unoptimized program does not trap; the
// original modify access may be on an unreachable code path. This is acceptable
// because:
//
// (a) in theory, exclusivity violations do not need to be executed to be
// considered program violations. Promoting the access does not introduce any
// new conflict where one didn't already exist statically. Catching these
// violations at runtime is only an implementation compromise, and the more true
// violations are caught, the better.
//
// (b) in practice, this situation is so exceedingly unlikely that it won't
// cause any pervasive usability problem where programs have stronger
// enforcement only when optimized.
```

In my mind that the bigger concern with hoisting traps early is that we trap before the program exits, but a postdominance check won't catch that. Note that we already do this for bounds checks, where I think the hoisting is a legitimate problem. With exclusivity violations I don't think it's a problem because an exclusivity violation is a static compiler error regardless of whether the code will be executed.